syntax = "proto2";

package ChakraProtoMsg;

enum NodeType
{
  INVALID_NODE = 0;
  METADATA_NODE = 1;
  MEM_LOAD_NODE = 2;
  MEM_STORE_NODE = 3;
  COMP_NODE = 4;
  COMM_SEND_NODE = 5;
  COMM_RECV_NODE = 6;
  COMM_COLL_NODE = 7;
}

enum AttributeType
{
  BOOL = 0;
  FLOAT = 1;
  UINT = 2;
  INT = 3;
  STRING = 4;
  BOOLS = 5;
  FLOATS = 6;
  UINTS = 7;
  INTS = 8;
  STRINGS = 9;
}

enum CollectiveCommType
{
  ALL_REDUCE = 0;
  ALL_TO_ALL = 1;
  ALL_GATHER = 2;
  REDUCE_SCATTER = 3;
}

message AttributeProto {
  required string name = 1;
  required AttributeType type = 2;
  optional string doc_string = 3;
  optional bool b = 4;
  optional float f = 5;
  optional uint64 u = 6;
  optional int64 i = 7;
  optional string s = 8;
  repeated bool bools = 9;
  repeated float floats = 10;
  repeated uint64 uints = 11;
  repeated int64 ints = 12;
  repeated string strings = 13;
}

message GlobalMetadata {
  repeated AttributeProto attribute = 1;
}

message Node {
  required uint64 id = 1;
  required string name = 2;
  required NodeType type = 3 [default = INVALID_NODE];
  repeated uint64 parent = 4;
  optional string inputs = 5;
  optional string input_shapes = 6;
  optional string input_types = 7;
  optional string outputs = 8;
  optional string output_shapes = 9;
  optional string output_types = 10;
  repeated AttributeProto attribute = 11;
}
